### PR TITLE
feat: [CI-5185]: RunTests with Parallelism

### PR DESCRIPTION
**Pull request by rutvijmehta-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37092)

File: product/ci/addon/tasks/runtests_test.go [hunk 1]
```diff
@@ -29,81 +27,81 @@ import (
...
-	fs.EXPECT().MkdirAll(expDir, os.ModePerm).Return(nil)
-	mf := filesystem.NewMockFile(ctrl)
-	mf.EXPECT().Write([]byte(expData)).Return(0, errors.New("could not write data"))
-	fs.EXPECT().Create("/test/tmp/config.ini").Return(mf, nil)
-
-	runner := mocks.NewMockTestRunner(ctrl)
-	_, err := r.createJavaAgentConfigFile(runner)
-	assert.NotNil(t, err)
-}
+//func TestCreateJavaAgentArg(t *testing.T) {
```
Comment (jamie-harness) [inline product/ci/addon/tasks/runtests_test.go:30]: why is this one commented out?

Comment (rutvijmehta-harness) [reply product/ci/addon/tasks/runtests_test.go:30]: That's because I need to rerun mockgen and fix the test issues. I'll address this before merging.

Comment (rutvijmehta-harness) [reply product/ci/addon/tasks/runtests_test.go]: Done

File: product/ci/addon/tasks/runtests.go [hunk 2]
```diff
@@ -377,7 +546,14 @@ func (r *runTestsTask) getCmd(ctx context.Context, agentPath, outputVarFile stri
 		}
 	}
 
-	testCmd, err := runner.GetCmd(ctx, selection.Tests, r.args, iniFilePath, isManual, !r.runOnlySelectedTests)
+	// Test splitting: only when parallelism is enabled
+	doNotSplit := false // Take from config as per design. Rename after review
+	if isParallelismEnabled() {
+		ignoreInstr = r.invokeParallelism(ctx, runner, &selection, ignoreInstr, doNotSplit)
```
Comment (hemanthmantri) [inline product/ci/addon/tasks/runtests.go:552]: This interface isn't intuitive/clean. invokeParallelism returning whether to instrument.

**Review by hemanthmantri — COMMENTED** [review]
First round of comments. Because this is a significant change, please describe it in the PR description message. Also, please make sure we have addressed the following:

- Running parallel tests in multiple steps of same stage is okay (with Filesystem sharing)
- No regression for .NET even if parallelism is enabled
- Parallel runs work with Jacoco agent attached - in same stage, to make sure it doesn't mess with directory structure
- The XML test reports generated are unique for each step so they aren't stepped over by each of the parallel stages.
- Total time computation is correct in the tests page
- How is the time saved computation going to work: Can we do `max(cpu time across all steps/stages) - sum(cpu time across all stages when a full run was invoked)`.

**Review by rutvijmehta-harness — COMMENTED** [review]
Running parallel tests in multiple steps of same stage is okay (with Filesystem sharing)
- Currently testing in Maven, figuring out the arguments which need to be tweaked. We'll setup a golden pipeline for all runners and document it so that users know what needs to be done.
- 
No regression for .NET even if parallelism is enabled
- Since the runner is not implemented, all tests will run in container 0.

Parallel runs work with Jacoco agent attached - in same stage, to make sure it doesn't mess with directory structure
- Under testing

The XML test reports generated are unique for each step so they aren't stepped over by each of the parallel stages.
- The user will have to configure test output using the parallel variables and set them in report paths.

Total time computation is correct in the tests page
- Ticket is filed for this

How is the time saved computation going to work: Can we do max(cpu time across all steps/stages) - sum(cpu time across all stages when a full run was invoked).
- Will file a ticket for this

**Review by hemanthmantri — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)* status: 17 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`product/ci/addon/tasks/runtests.go` line 40 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGYtOJesKdQaSLp-r2-40:-NBxHkLM8sBXHWecd0om:b2nhb2c) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/tasks/runtests.go#L40)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Ack
___
*[`product/ci/addon/tasks/runtests.go` line 96 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGYtOJesKdQaSLp-r2-96:-NBxHt2Y4GCJZccNjIqp:b-b6hioz) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/tasks/runtests.go#L96)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

Enum is not natively supported in Go. We use string for other attrs like Language, Build Tool, etc.
</blockquote></details>

https://go.dev/play/p/HhAy3YlI6_q seems to have an example of how to use enums. If we know the set of expected values, we should prefer enums over arbitrary strings. You can choose to defer this change.
___
*[`product/ci/addon/tasks/runtests.go` line 332 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGYtOJesKdQaSLp-r2-332:-NBxIKqh2b2QNV1YeE1c:b-59tomz) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/tasks/runtests.go#L332)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

Done
</blockquote></details>

allTests
___
*[`product/ci/addon/tasks/runtests.go` line 552 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGYtOJesKdQaSLp-r2-552:-NBxItsQAGY8oixVYjiq:b-u13une) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/tasks/runtests.go#L552)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

Done
</blockquote></details>

The name is changed but it is still returning ignoreInstr -> Seems unclean to me. 
___
*[`product/ci/addon/testintelligence/java/gradle.go` line 44 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGZFd1JX9W4Yk_w-r2-44:-NBxJAO01om4oYRbO36D:b71n44z) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/testintelligence/java/gradle.go#L44)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

It's already consistent. Both the methods start with Auto.
</blockquote></details>

Got it. The folded view showed only `return DetectPkgs` so I missed the actual method name.
___
*[`product/ci/addon/testintelligence/java/helper.go` line 65 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGZFd1JX9W4Yk_x-r2-65:-NBxJMQX7sFRn4-zbsMY:b-k18lfe) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/testintelligence/java/helper.go#L65)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

Both are separate code paths and one can exist without the other. Let's think more on this.
</blockquote></details>

Yes, the code can be common. The most common use case should be optimized (which should be fully auto)
___
*[`product/ci/split_tests/utils/consts.go` line 18 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NBQhjQux5w8hspudqVy-r2-18:-NBxJyYI7FaHCAjCy03T:b-swc37z) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/split_tests/utils/consts.go#L18)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

Done
</blockquote></details>

SplitByTestCount

**Comment by rutvijmehta-harness** [comment]
*[`product/ci/addon/tasks/runtests.go` line 552 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGYtOJesKdQaSLp-r2-552:-NBxSWF78GlMyc3oi9Q-:b-um7t64) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/tasks/runtests.go#L552)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

The name is changed but it is still returning ignoreInstr -> Seems unclean to me. 
</blockquote></details>

This method is supposed to update the ignoreInstr flag. If it's a manual run, we do not want to run all the tests (check runner code and also TODO in computeSelectedTests.

**Comment by rutvijmehta-harness** [comment]
*[`product/ci/addon/testintelligence/java/helper.go` line 65 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGZFd1JX9W4Yk_x-r2-65:-NBxT1cd5RGS0ffChZ8c:b-qg0siu) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/testintelligence/java/helper.go#L65)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

Yes, the code can be common. The most common use case should be optimized (which should be fully auto)
</blockquote></details>

It's not common code as well. DetectPkgs is reading through file data and DetectJavaTests is processing the file path as a string.

**Comment by rutvijmehta-harness** [comment]
*[`product/ci/split_tests/utils/consts.go` line 18 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NBQhjQux5w8hspudqVy-r2-18:-NBxTP7SAr3EdVITRK3g:bb6sh2i) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/split_tests/utils/consts.go#L18)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

SplitByTestCount
</blockquote></details>

Ack.

**Comment by rutvijmehta-harness** [comment]
*[`product/ci/addon/tasks/runtests.go` line 332 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGYtOJesKdQaSLp-r2-332:-NBxTToTD7dhoH16xXqE:b-ahd48) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/tasks/runtests.go#L332)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

allTests
</blockquote></details>

allTests is misleading. I'll switch to testsToSplit if tests is not okay

**Comment by rutvijmehta-harness** [comment]
*[`product/ci/addon/tasks/runtests.go` line 96 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGYtOJesKdQaSLp-r2-96:-NBxTf-ZDsZUf5esiJa_:bjvm8ch) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/tasks/runtests.go#L96)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

https://go.dev/play/p/HhAy3YlI6_q seems to have an example of how to use enums. If we know the set of expected values, we should prefer enums over arbitrary strings. You can choose to defer this change.
</blockquote></details>

Java (CI Manager) will be Enum which will send the split strategy as a string (converted from enum). Hence there is very little chance of us getting an unknown string. If the string is unknown, we have defaults in our switch cases as a safety net. This can be an enhancement to convert all attributes to a enums by implementing individual types.

**Comment by rutvijmehta-harness** [comment]
*[`product/ci/addon/tasks/runtests.go` line 452 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGYtOJesKdQaSLp-r2-452:-NBxV76R63rv-vJBbrp-:b1q6ez2) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/tasks/runtests.go#L452)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

It will require some kind of string handling. I can look more into this.
</blockquote></details>

Looked into this, string handling will require couple of if else cases to handle scenarios where stage and step parallelism is enabled/disabled. We can target this in upcoming PR.

**Comment by rutvijmehta-harness** [comment]
*[`product/ci/addon/tasks/runtests.go` line 422 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGYtOJesKdQaSLp-r2-422:-NBxhA1B3zSf5eSUkSxV:bvros29) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/tasks/runtests.go#L422)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

These should not be needed. If you had to do this please reopen https://harness.atlassian.net/browse/PIE-4769 and check with Sahil.
</blockquote></details>

Following up in slack conversation.

File: product/ci/addon/tasks/runtests.go [hunk 3]
```diff
@@ -312,6 +331,153 @@ func (r *runTestsTask) getTestSelection(ctx context.Context, files []types.File,
...
+	currentTestSet := make(map[string]bool)
+	var testID string
+	for _, t := range testsToSplit {
+		switch splitStrategy {
+		case classTimingTestSplitStrategy:
+			testID = t.Pkg + t.Class
+		case countTestSplitStrategy:
+			testID = t.Pkg + t.Class
+		default:
+			testID = t.Pkg + t.Class
```
Comment (hemanthmantri) [inline product/ci/addon/tasks/runtests.go:351]: All cases have the same testID. What am I missing?

File: product/ci/addon/tasks/runtests.go [hunk 4]
```diff
@@ -377,7 +548,13 @@ func (r *runTestsTask) getCmd(ctx context.Context, agentPath, outputVarFile stri
 		}
 	}
 
-	testCmd, err := runner.GetCmd(ctx, selection.Tests, r.args, iniFilePath, isManual, !r.runOnlySelectedTests)
+	// Test splitting: only when parallelism is enabled
+	if isParallelismEnabled() {
+		ignoreInstr = r.computeSelectedTests(ctx, runner, &selection, ignoreInstr)
```
Comment (hemanthmantri) [inline product/ci/addon/tasks/runtests.go:553]: Still have the same comment about this function's interface. If you have to change the ignoreInstr anyway can you pass &ignoreInstr (feels less ugly).

**Review by hemanthmantri — APPROVED** [review]

File: product/ci/addon/tasks/runtests.go (hunk 3 above)
Comment (rutvijmehta-harness) [reply product/ci/addon/tasks/runtests.go:351]: I initially had an additional switch case for file path where one case was different. We can merge the cases but I'd still like to keep the switch case.

File: product/ci/addon/tasks/runtests.go (hunk 4 above)
Comment (rutvijmehta-harness) [reply product/ci/addon/tasks/runtests.go:553]: Passed by reference.

**Review by hemanthmantri — DISMISSED** [review]

**Review by jamie-harness — COMMENTED** [review]
> Reviewed 7 of 16 files at r1, 5 of 8 files at r2, 2 of 5 files at r4, 1 of 3 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)* status: 10. unresolved discussions. 0 of 1 approvals obtained.

**Review by jamie-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)* status: 9. unresolved discussions.

**Review by hemanthmantri — COMMENTED** [review]
> Reviewed 6 of 16 files at r1, 5 of 8 files at r2, 2 of 5 files at r4, 1 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)* status: 3. unresolved discussions. 

___
*[`product/ci/addon/testintelligence/java/helper.go` line 65 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGZFd1JX9W4Yk_x-r2-65:-NC1NDQL2d7GOICG_3tM:bd88dr4) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/testintelligence/java/helper.go#L65)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

It's not common code as well. DetectPkgs is reading through file data and DetectJavaTests is processing the file path as a string.
</blockquote></details>

Let's file a ticket to make sure we detect pkgs and tests in one FS walk.

**Comment by rutvijmehta-harness** [comment]
*[`product/ci/addon/testintelligence/java/helper.go` line 65 at r2](https://reviewable.io/reviews/harness/harness-core/37092#-NB-ymGZFd1JX9W4Yk_x-r2-65:-NC1Qsfl9FZoQoC5JjFn:b-gr9jgp) ([raw file](https://github.com/harness/harness-core/blob/62e90fe31c6d5058cbaf493d75458d013e6c1d35/product/ci/addon/testintelligence/java/helper.go#L65)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

Let's file a ticket to make sure we detect pkgs and tests in one FS walk.
</blockquote></details>

Discussed offline. One method is reading through the files and other one is working on file paths. So merging this 2 methods is not needed.

**Review by hemanthmantri — APPROVED** [review]
:lgtm:

> Reviewed 1 of 16 files at r1, 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092#-:-NC1S-pR1Ry15bJ-YLiY:bnfp4nl)* status: :shipit: complete! ✓

**Review by hemanthmantri — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)* status: :shipit: complete! ✓

**Review by hemanthmantri — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)* status: :shipit: complete! ✓

**Review by hemanthmantri — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)* status: :shipit: complete! ✓

**Review by dgarg-harness — APPROVED** [review]

**Review by USB-05 — APPROVED** [review]

**Review by USB-05 — DISMISSED** [review]
> Reviewed 7 of 16 files at r1, 5 of 8 files at r2, 2 of 5 files at r4, 1 of 3 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)* status: Some of the checks are too old.

**Review by diptiman-adak — DISMISSED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37092)* status: Some of the checks are too old.

**Comment by rutvijmehta-harness** [comment]
Rebased with develop for recency check to pass.

**Pull request merged**